src/dawn/Optimizer/PassSetBoundaryCondition.cpp

```diff
@@ -0,0 +1,332 @@
+      stencilCallsToReplace_.emplace_back(stmt);
+  }
+
+  std::vector<std::shared_ptr<StencilCallDeclStmt>>& getStencilCallsToReplace() {
+    return stencilCallsToReplace_;
+  }
+
+  void reset() { stencilCallsToReplace_.clear(); }
+};
+
+PassSetBoundaryCondition::PassSetBoundaryCondition() : Pass("PassSetBoundaryCondition") {}
+
+bool PassSetBoundaryCondition::run(StencilInstantiation* stencilInstantiation) {
+
+  // check if we need to run this pass
+  if(stencilInstantiation->getStencils().size() == 1) {
+    if(stencilInstantiation->getOptimizerContext()->getOptions().ReportBoundaryConditions) {
+      std::cout << "\nPASS: " << getName() << ": " << stencilInstantiation->getName() << " :";
+      std::cout << " No boundary conditions applied\n";
+    }
+    return true;
+  }
+  // returns the original ID of a variable
+  auto getOriginalID = [&](int ID) {
+    // check if the variable exists to be sure not to assert
+    auto it = stencilInstantiation->getNameToAccessIDMap().find(
+        stencilInstantiation->getOriginalNameFromAccessID(ID));
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      if(stencilInstantiation->isField(ID)) {
+        return stencilInstantiation->getAccessIDFromName(
+            stencilInstantiation->getOriginalNameFromAccessID(ID));
+      } else {
+        return (int) FieldType::FT_NotOriginal;
+      }
+    } else {
+      return (int) FieldType::FT_NotOriginal;
+    }
+
+  };
+
+  std::unordered_map<int, Extents> dirtyFields;
+  std::unordered_map<int, std::shared_ptr<BoundaryConditionDeclStmt>> allBCs;
+
+  // Fetch all the boundary conditions stored in the instantiation
+  for(const auto& bc : stencilInstantiation->getBoundaryConditions()) {
+    auto it = stencilInstantiation->getNameToAccessIDMap().find(bc.first);
+    if(it != stencilInstantiation->getNameToAccessIDMap().end()) {
+      int ID = stencilInstantiation->getAccessIDFromName(bc.first);
+      allBCs.emplace(ID, bc.second);
+    }
+  }
+
+  // Get the order in which the stencils are called:
+  VisitStencilCalls findStencilCalls;
+
+  for(const std::shared_ptr<Statement>& statement :
+      stencilInstantiation->getStencilDescStatements()) {
+    statement->ASTStmt->accept(findStencilCalls);
+  }
+  std::vector<int> StencilIDsVisited_;
+  for(const auto& stencilcall : findStencilCalls.getStencilCalls()) {
+    StencilIDsVisited_.push_back(
+        stencilInstantiation->getStencilCallToStencilIDMap().find(stencilcall)->second);
+  }
+
+  auto calculateHaloExtents = [&](std::string fieldname) {
+
+    Extents fullExtent;
+    // Did we already apply a BoundaryCondition for this field?
+    // This is the first time we apply a BC to this field, we traverse all stencils that were
+    // applied before
+    std::vector<int> stencilIDsToVisit(StencilIDsVisited_);
+    if(StencilBCsApplied_.count(fieldname) == 0) {
+    } else {
+      for(int traveresedID : StencilBCsApplied_.find(fieldname)->second) {
```

Comment:
assert that the key exists

Reply:
StencilBCsApplied_ is only inserted by this lambda function, apparently, later at the end. 
So is it ever possible that contains elements for field key fieldname here? It does not look like this is the case

Reply:
no need to assert, we just do not do the erase (the empty if)